src/main/kotlin/koren/proj/puzzlesolver/model/searchAlgo/DFS.kt

```diff
@@ -1,13 +1,51 @@
 package koren.proj.puzzlesolver.model.searchAlgo
 
 import koren.proj.puzzlesolver.model.puzzle.AbstractPuzzle
+import java.util.*
+import kotlin.collections.HashMap
+import kotlin.collections.HashSet
 
 class DFS : GenericSearchInterface {
 
-    override var parentMap: HashMap<AbstractPuzzle, AbstractPuzzle> = HashMap()
-    override var visitedSet: HashSet<AbstractPuzzle> = HashSet()
+    private val parentMap: HashMap<AbstractPuzzle, AbstractPuzzle> = HashMap()
+    private val visitedSet: HashSet<AbstractPuzzle> = HashSet()
 
-    override fun search(initial_puzzle: AbstractPuzzle, goal: AbstractPuzzle) {
-        TODO("Not yet implemented")
+    override fun search(initialPuzzle: AbstractPuzzle, goal: AbstractPuzzle): Collection<AbstractPuzzle> {
+        var currentPuzzle = initialPuzzle
+        val searchQueue: Stack<AbstractPuzzle> = Stack<AbstractPuzzle>()
+        searchQueue.push(currentPuzzle)
+
+        while (!searchQueue.isEmpty() && currentPuzzle != goal) {
+            currentPuzzle = searchQueue.pop()
+            if (!visitedSet.contains(currentPuzzle)) {
+                visitedSet.add(currentPuzzle)
+                val currentNeighbors = currentPuzzle.generateSteps()
+
+                for (currentNeighbor in currentNeighbors) {
+                    if (!visitedSet.contains(currentNeighbor)) {
+                        parentMap[currentNeighbor] = currentPuzzle
+                        searchQueue.push(currentNeighbor)
+                    }
+                }
+            }
+        }
+
+        return if (currentPuzzle == goal) {
+            inferStepsToGoal(initialPuzzle, goal)
+        } else {
+            LinkedList<AbstractPuzzle>()
```

Comment:
Think about how to differentiate between a search which has no solution, and a search which the solution is not to move

Reply:
Adding test for this case